akka-serialization-jackson/src/test/scala/akka/serialization/jackson/JacksonSerializerSpec.scala

```diff
@@ -626,6 +628,20 @@ abstract class JacksonSerializerSpec(serializerName: String)
     val manifest = serializer.manifest(obj)
     val serializerId = serializer.identifier
     val blob = serializeToBinary(obj)
+
+    // Issue #28918, check that CBOR format is used (not JSON).
+    if (blob.length > 0) {
+      serializer match {
+        case _: JacksonJsonSerializer =>
+          if (!JacksonSerializer.isGZipped(blob))
+            new String(blob.take(1), StandardCharsets.UTF_8) should ===("{")
+        case _: JacksonCborSerializer =>
+          new String(blob.take(1), StandardCharsets.UTF_8) should !==("{")
+        case _ =>
+          throw new IllegalArgumentException(s"Unexpected serializer $serializer")
+      }
+    }
+
     val deserialized = deserializeFromBinary(blob, serializerId, manifest, sys)
     deserialized should ===(obj)
   }
```

Comment:
Let's discuss how we should support compatibility (rolling updates, persistent data). I think it requires a two phase update.

```
  serializers {
    jackson-json = "akka.serialization.jackson.JacksonJsonSerializer"
    jackson-cbor = "akka.serialization.jackson.JacksonCborSerializer"
  }
  serialization-identifiers {
    jackson-json = 31
    jackson-cbor = 32
  }
```

For Akka 2.6.5:
* First step could be to bind 32 to JacksonJsonSerializer, since that is essentially what it does today. toBinary will continue to be JSON, which can still be read by JacksonCborSerializer in 2.6.4.
* Change JacksonCborSerializer so that it use the correct CBORFactory and bind it to new serializerId 33. This will not be used for toBinary yet.

For Akka 2.6.6:
* Second step would be to bind 33 jackson-cbor JacksonCborSerializer also for toBinary.
* Old data is still with serializer 32 and will be bound to JacksonJsonSerializer for fromBinary.

Reply:
For compatibility... it'll be interesting to see how CRDTs containing CBOR will respond to this (got alot of these in Cloudstate management API) - the serialized version (and hence hash) will change, while the actual data won't, but two nodes on two different versions will inifinitely replicate everything to each other. Hopefully that will just sort itself out once all nodes are upgraded.

Another thing, you could implement detection of cbor or json in decoding (and make it configurable). If we assume that every message is either an object or array when serialized, then there's a very simple heuristic, if the first bit is 1, it's CBOR, if it's 0, it's JSON (CBOR objects start with 0b101, and arrays start with 0b100, while JSON always starts with an ascii character so the first bit is always 0). This should work in close to 100% of real world use cases, when would you ever use Jackson to make a message serialize to a number, string, boolean or null? You wouldn't, you'd use a built in Akka serializer.

Reply:
> it'll be interesting to see how CRDTs containing CBOR will respond 

In the second rolling update (when changing to real cbor) it will sync over and over with the old nodes. It shouldn't be more than once per node per gossip interval so not much traffic unless you have huge amount of data. It will stop when all old nodes have been phased out.